pkg/util/cache/intern.go

```diff
@@ -0,0 +1,359 @@
+	return i
+}
+
+// loadOrStore always returns the string from the cache, adding it into the
+// cache if needed.
+// If we need to store a new entry and the cache is at its maximum capacity,
+// it is reset.  The origin identifies the container (and thusly, the quota)
+// originating the string.
+func (i *stringInterner) loadOrStore(key []byte) string {
+	if len(key) == 0 {
+		// Empty key
+		return ""
+	}
+	if i.maxStringCount == 0 {
+		// Dead interner.  release() has already been called, or it was broken on
+		// construction.
+		return ""
+	}
+
+	result := i.cache.lookupOrInsert(key, func(key []byte) string {
+		if i.fileBacking != nil {
+			s, possible := i.fileBacking.lookupOrInsert(key)
+			if len(s) == 0 {
+				if !possible {
+					// String is too large to allocate in the backing, just do it
+					// on the heap.  Let GC take care of it normally.  The string cache
+					// already de-duplicates for us.
+					return string(key)
+				}
+				// Return the empty string here, and "loadOrStore's" caller will
+				// resize the interner in response.
+				return ""
+			}
+			return s
+		}
+		return string(key)
+	})
+
+	return result
+}
+
+func (i *stringInterner) used() int64 {
+	used, _ := i.fileBacking.sizes()
+	return used
+}
+
+// Retain some references
+func (i *stringInterner) Retain(n int32) {
+	i.refcountLock.Lock()
+	defer i.refcountLock.Unlock()
+	i.refcount += n
+}
+
+// Release some references
+func (i *stringInterner) Release(n int32) {
+	i.refcountLock.Lock()
+	defer i.refcountLock.Unlock()
+	if i.refcount < 1 {
+		log.Warnf("Dead stringInterner being released!  refcount=%d", i.refcount)
+		return
+	}
+	if i.refcount > 0 && i.refcount-n < 1 {
+		log.Debugf("Finalizing backing, refcount=%d, n=%d", i.refcount, n)
+		i.fileBacking.finalize()
+		i.cache = newLruStringCache(0, false)
+		i.maxStringCount = 0
+	}
+	i.refcount -= n
+}
+
+func (i *stringInterner) retain() {
+	i.refcountLock.Lock()
+	defer i.refcountLock.Unlock()
+	if i.refcount < 1 {
+		log.Error("Dead interner being re-retained!")
+	}
+	i.refcount++
+}
+
+// KeyedInterner has an origin-keyed set of interners.
+type KeyedInterner struct {
+	interners      *lru.Cache
+	maxQuota       int
+	closeOnRelease bool
+	tmpPath        string
+	lastReport     time.Time
+	minFileSize    int64
+	maxPerInterner int
+	lock           sync.Mutex
+}
+
+// NewKeyedStringInterner creates a Keyed String Interner with a max per-origin quota of maxQuota
+func NewKeyedStringInterner(cfg cconfig.Component) *KeyedInterner {
+	stringInternerCacheSize := cfg.GetInt("dogstatsd_string_interner_size")
+	enableMMap := cfg.GetBool("dogstatsd_string_interner_mmap_enable")
+
+	if enableMMap {
+		closeOnRelease := !cfg.GetBool("dogstatsd_string_interner_mmap_preserve")
+		tempPath := cfg.GetString("dogstatsd_string_interner_tmpdir")
+		minSizeKb := cfg.GetInt("dogstatsd_string_interner_mmap_minsizekb")
+		maxStringsPerInterner := cfg.GetInt("dogstatsd_string_interner_origin_max_strings")
```

Comment:
```suggestion
		maxStringsPerInterner := cfg.GetInt("dogstatsd_string_interner_per_origin_max_strings")
```

Reply:
Renaming to `dogstatsd_string_interner_per_origin_initial_size` to both integrate your commend and better describe the config parameter.

Reply:
Ok, new entries in `config.go` should all have inline docs at declaration now.